src/orchestratord/src/controller/console.rs

```diff
@@ -0,0 +1,544 @@
+    },
+};
+use kube::{
+    Api, Client, Resource, ResourceExt,
+    api::{ObjectMeta, PostParams},
+    runtime::{
+        controller::Action,
+        reflector::{ObjectRef, Store},
+    },
+};
+use maplit::btreemap;
+use serde::Serialize;
+use tracing::trace;
+
+use crate::{
+    Error,
+    k8s::{apply_resource, make_reflector},
+    tls::{DefaultCertificateSpecs, create_certificate, issuer_ref_defined},
+};
+use mz_cloud_resources::crd::{
+    ManagedResource,
+    console::v1alpha1::Console,
+    generated::cert_manager::certificates::{Certificate, CertificatePrivateKeyAlgorithm},
+};
+use mz_orchestrator_kubernetes::KubernetesImagePullPolicy;
+use mz_ore::{cli::KeyValueArg, instrument};
+use mz_server_core::listeners::AuthenticatorKind;
+
+pub struct Config {
+    pub enable_security_context: bool,
+    pub enable_prometheus_scrape_annotations: bool,
+
+    pub image_pull_policy: KubernetesImagePullPolicy,
+    pub scheduler_name: Option<String>,
+    pub console_node_selector: Vec<KeyValueArg<String, String>>,
+    pub console_affinity: Option<Affinity>,
+    pub console_tolerations: Option<Vec<Toleration>>,
+    pub console_default_resources: Option<ResourceRequirements>,
+    pub network_policies_ingress_enabled: bool,
+    pub network_policies_ingress_cidrs: Vec<String>,
+
+    pub default_certificate_specs: DefaultCertificateSpecs,
+
+    pub console_http_port: u16,
+    pub balancerd_http_port: u16,
+}
+
+#[derive(Serialize)]
+struct AppConfig {
+    version: String,
+    auth: AppConfigAuth,
+}
+
+#[derive(Serialize)]
+struct AppConfigAuth {
+    mode: AuthenticatorKind,
+}
+
+pub struct Context {
+    config: Config,
+    deployments: Store<Deployment>,
+}
+
+impl Context {
+    pub async fn new(config: Config, client: Client) -> Self {
+        Self {
+            config,
+            deployments: make_reflector(client).await,
+        }
+    }
+
+    async fn sync_deployment_status(
+        &self,
+        client: &Client,
+        console: &Console,
+    ) -> Result<(), kube::Error> {
+        let namespace = console.namespace().unwrap();
+        let console_api: Api<Console> = Api::namespaced(client.clone(), &namespace);
+
+        let Some(deployment) = self
+            .deployments
+            .get(&ObjectRef::new(&console.deployment_name()).within(&namespace))
+        else {
+            return Ok(());
+        };
+
+        let Some(deployment_conditions) = &deployment
+            .status
+            .as_ref()
+            .and_then(|status| status.conditions.as_ref())
+        else {
+            // if the deployment doesn't have any conditions set yet, there
+            // is nothing to sync
+            return Ok(());
+        };
+
+        let ready = deployment_conditions
+            .iter()
+            .any(|condition| condition.type_ == "Available" && condition.status == "True");
+        let ready_str = if ready { "True" } else { "False" };
+
+        let mut status = console.status.clone().unwrap();
```

Comment:
Are we guaranteed to have a status at this point? Can we change to `expect` with an explanation of where we have ensured we have one?